profile/src/main/java/rdx/works/profile/data/repository/ProfileRepository.kt

```diff
@@ -54,113 +50,146 @@ suspend fun ProfileRepository.updateProfile(updateAction: suspend (Profile) -> P
+    private val profileStateChangeDriver: AndroidProfileStateChangeDriver,
     @IoDispatcher private val ioDispatcher: CoroutineDispatcher,
     @ApplicationScope applicationScope: CoroutineScope
 ) : ProfileRepository {
 
     init {
         applicationScope.launch {
-            val snapshotResult = encryptedPreferencesManager.encryptedProfile.firstOrNull()
-
-            if (snapshotResult == null) {
-                profileStateFlow.update { ProfileState.None }
-            } else {
-                snapshotResult
-                    .onFailure { exception ->
-                        logNonFatalException(exception)
-                        profileStateFlow.update { ProfileState.Incompatible(cause = exception) }
+            // Observe profile state changes that need to be backed up
+            profileStateChangeDriver
+                .profileState
+                .runningFold(
+                    initial = ProfileStateChange.initial(),
+                    operation = { history, profileState ->
+                        history.updateChanges(profileState)
                     }
-                    .onSuccess { snapshot ->
-                        ensureP2PLinkMigrationAcknowledged(snapshot)
-                        profileStateFlow.update { deriveProfileState(snapshot) }
+                )
+                .mapNotNull { it.updatedProfileForBackup() }
+                .collect { profile ->
+                    if (profile.canBackupToCloud) {
+                        cloudBackupSyncExecutor.requestCloudBackup()
+                    } else {
+                        encryptedPreferencesManager.clearProfileSnapshotFromCloudBackup()
                     }
-            }
+                }
         }
     }
 
-    private val profileStateFlow: MutableStateFlow<ProfileState> = MutableStateFlow(ProfileState.NotInitialised)
-
-    override val profileState = profileStateFlow
-        .filterNot { it is ProfileState.NotInitialised }
+    override val profileState = profileStateChangeDriver
+        .profileState
+        // Waits until the profile state is evaluated
+        .filterNotNull()
 
     override val inMemoryProfileOrNull: Profile?
-        get() = when (val state = profileStateFlow.value) {
-            is ProfileState.Restored -> state.profile
+        get() = when (val state = profileStateChangeDriver.profileState.value) {
+            is ProfileState.Loaded -> state.v1
             else -> null
         }
 
     override suspend fun saveProfile(profile: Profile) {
-        val hostId = hostInfoRepository.getHostId()
-        val hostInfo = hostInfoRepository.getHostInfo()
+        val sargonOs = sargonOsManager.sargonOs
+        val hostId = hostInfoRepository.getHostId().getOrNull() ?: return
+        val hostInfo = hostInfoRepository.getHostInfo().getOrNull() ?: return
+
         val profileToSave = profile.copy(
             header = profile.header.copy(
-                lastModified = TimestampGenerator(),
```

Comment:
I assume you removed this because it is generated in Sargon?

Reply:
Exactly, on each `set_profile` sargon modifies the timestamp. No need for the host to do the same.